client/go/outline/vpn/nmconn_linux.go

```diff
@@ -0,0 +1,188 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpn
+
+import (
+	"encoding/binary"
+	"log/slog"
+	"net"
+	"time"
+
+	gonm "github.com/Wifx/gonetworkmanager/v2"
+	"golang.org/x/sys/unix"
+)
+
+type nmConnectionOptions struct {
+	Name            string
+	TUNName         string
+	TUNAddr4        net.IP
+	DNSServers4     []net.IP
+	FWMark          uint32
+	RoutingTable    uint32
+	RoutingPriority uint32
+}
+
+func (c *linuxVPNConn) establishNMConnection() (err error) {
+	defer func() {
+		if err != nil {
+			c.closeNMConnection()
+		}
+	}()
+
+	if c.nm, err = gonm.NewNetworkManager(); err != nil {
+		return errSetupVPN(nmLogPfx, "failed to connect", err)
+	}
+	slog.Debug(nmLogPfx + "connected")
+
+	dev, err := c.waitForDeviceToBeAvailable()
+	if err != nil {
+		return errSetupVPN(nmLogPfx, "failed to find TUN device", err, "tun", c.nmOpts.TUNName)
+	}
+	slog.Debug(nmLogPfx+"located TUN device", "tun", c.nmOpts.TUNName, "dev", dev.GetPath())
+
+	if err = dev.SetPropertyManaged(true); err != nil {
+		return errSetupVPN(nmLogPfx, "failed to set TUN device to be managed", err, "dev", dev.GetPath())
+	}
+	slog.Debug(nmLogPfx+"set TUN device to be managed", "dev", dev.GetPath())
+
+	props := make(map[string]map[string]interface{})
+	configureCommonProps(props, c.nmOpts)
+	configureTUNProps(props)
+	configureIPv4Props(props, c.nmOpts)
+	slog.Debug(nmLogPfx+"populated NetworkManager connection settings", "settings", props)
+
+	// The previous SetPropertyManaged call needs some time to take effect
+	for retries := 10; retries > 0; retries-- {
+		slog.Debug(nmLogPfx+"trying to create connection for TUN device ...", "dev", dev.GetPath())
+		c.ac, err = c.nm.AddAndActivateConnection(props, dev)
+		if err == nil {
+			break
+		}
+		slog.Debug(nmLogPfx+"waiting for TUN device being managed", "err", err)
+		time.Sleep(50 * time.Millisecond)
+	}
+	if err != nil {
+		return errSetupVPN(nmLogPfx, "failed to create new connection for device", err, "dev", dev.GetPath())
+	}
+	slog.Info(nmLogPfx+"successfully configured NetworkManager connection", "conn", c.ac.GetPath())
+	return nil
+}
+
+func (c *linuxVPNConn) closeNMConnection() error {
+	if c == nil || c.nm == nil {
+		return nil
+	}
+
+	if c.ac != nil {
+		if err := c.nm.DeactivateConnection(c.ac); err != nil {
+			slog.Warn(nmLogPfx+"not able to deactivate connection", "err", err, "conn", c.ac.GetPath())
+		}
+		slog.Debug(nmLogPfx+"deactivated connection", "conn", c.ac.GetPath())
+
+		conn, err := c.ac.GetPropertyConnection()
+		if err == nil {
+			err = conn.Delete()
+		}
+		if err != nil {
+			return errCloseVPN(nmLogPfx, "failed to delete connection", err, "conn", c.ac.GetPath())
+		}
+		slog.Info(nmLogPfx+"connection deleted", "conn", c.ac.GetPath())
+	}
+
+	return nil
+}
+
+func (c *linuxVPNConn) waitForDeviceToBeAvailable() (dev gonm.Device, err error) {
```

Comment:
I'd make this standalone and take the nm and TUNName instead. That will make it clearer what is involved in this operation.

Reply:
Done, moved to `tun_linux.go`.